core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregator.java

```diff
@@ -0,0 +1,234 @@
+    private final String separator;
+
+    public AdjacencyMatrixAggregator(String name, AggregatorFactories factories, String separator, String[] keys, Weight[] filters,
+            SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
+            throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        this.separator = separator;
+        this.keys = keys;
+        this.filters = filters;
+        this.totalNumIntersections = ((keys.length * keys.length) - keys.length) / 2;
+        this.totalNumKeys = keys.length + totalNumIntersections;
+    }
+
+    private static class BitsIntersector implements Bits {
+        Bits a;
+        Bits b;
+
+        public BitsIntersector(Bits a, Bits b) {
+            super();
+            this.a = a;
+            this.b = b;
+        }
+
+        @Override
+        public boolean get(int index) {
+            return a.get(index) && b.get(index);
+        }
+
+        @Override
+        public int length() {
+            return Math.min(a.length(), b.length());
+        }
+
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, final LeafBucketCollector sub) throws IOException {
+        // no need to provide deleted docs to the filter
+        final Bits[] bits = new Bits[filters.length + totalNumIntersections];
+        for (int i = 0; i < filters.length; ++i) {
+            bits[i] = Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), filters[i].scorer(ctx));
+        }
+        // Add extra Bits for intersections
+        int pos = filters.length;
+        for (int i = 0; i < filters.length; i++) {
+            for (int j = i + 1; j < filters.length; j++) {
+                bits[pos++] = new BitsIntersector(bits[i], bits[j]);
+            }
+        }
+        return new LeafBucketCollectorBase(sub, null) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                for (int i = 0; i < bits.length; i++) {
+                    if (bits[i].get(doc)) {
+                        collectBucket(sub, doc, bucketOrd(bucket, i));
+                    }
+                }
+            }
+        };
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long owningBucketOrdinal) throws IOException {
+
+        // Buckets are ordered into groups - [keyed filters] [key1&key2 intersects]
+
+        List<InternalAdjacencyMatrix.InternalBucket> buckets = new ArrayList<>(filters.length);
+        for (int i = 0; i < keys.length; i++) {
+            long bucketOrd = bucketOrd(owningBucketOrdinal, i);
+            int docCount = bucketDocCount(bucketOrd);
+            // Empty buckets are not returned because this aggregation will commonly be used under a
+            // a date-histogram where we will look for transactions over time and can expect many
+            // empty buckets.
```

Comment:
This worries me a bit as this is inconsistent with the filters and ranges aggregations.